torch/_refs/__init__.py

```diff
@@ -2944,6 +2951,108 @@ def unbind(t: TensorLikeType, dim: int = 0) -> TensorSequenceType:
     )
 
 
+@register_decomposition(torch.ops.aten.index_copy)
+@out_wrapper()
+def index_copy(x: TensorLike, dim: int, index: TensorLike, tensor: TensorLike):
+    return x.clone().index_copy_(dim, index, tensor)
+
+
+# The decomposition of this function dispatches to aten.index_put_ for efficiency
+# We cannot do that in Python, as torch.index_put_ does not support slice(None)s See
+# https://github.com/pytorch/pytorch/pull/85002#issuecomment-1248524492
+def index_copy_(x: TensorLike, dim: int, index: TensorLike, tensor: TensorLike):
```

Comment:
do you need a separate decomposition for this one? y[idx] translates into slice + index_put call without extra copies:
```
In [8]: def fn(x, y, z):
   ...:     idx = (slice(None),) + (z,)
   ...:     x[idx]=y
   ...: 

In [9]: z=torch.arange(3)

In [10]: with LoggingMode():
    ...:     fn(x,y,z)
    ...: 
$1: f32[4, 4] = aten.slice.Tensor($0, 0, 0, 9223372036854775807)
$1: f32[4, 4] = aten.index_put_.default($1, [None, $2], $3)
```

Reply:
This one we can simply dispatch to `index_put_` directly so we're saving one kernel, I guess. Not that it changes too much. My reasoning was to have separate decompositions for all the setters.

Reply:
`slice` is op on strides, so we don't save a kernel. It's better to not have separate decompositions if we can avoid it.

Reply:
That's fair. I'll roll that one back. I'm just leaving `index_add` as an independent decomposition.